src/libraries/System.Private.DataContractSerialization/src/System/Runtime/Serialization/DataContractSet.cs

```diff
@@ -197,88 +202,404 @@ private void AddKnownDataContracts(DataContractDictionary? knownDataContracts)
+                            referencedTypes.Add(xmlName, types);
+                        }
+                    }
+                    else
+                    {
+                        List<Type> types = (List<Type>)value;
+                        if (!types.Contains(type))
+                            types.Add(type);
+                    }
+                }
+                else
+                    referencedTypes.Add(xmlName, type);
+            }
+        }
+
+        [RequiresUnreferencedCode(DataContract.SerializerTrimmerWarning)]
+        private static bool IsTypeReferenceable(Type type)
+        {
+            try
+            {
+                return (type.IsSerializable ||
+                        type.IsDefined(Globals.TypeOfDataContractAttribute, false) ||
+                        (Globals.TypeOfIXmlSerializable.IsAssignableFrom(type) && !type.IsGenericTypeDefinition) ||
+                        CollectionDataContract.IsCollection(type, out _) ||
+                        ClassDataContract.IsNonAttributedTypeValidForSerialization(type));
+            }
+            catch (Exception)
+            {
+                // An exception can be thrown in the designer when a project has a runtime binding redirection for a referenced assembly or a reference dependent assembly.
+                // Type.IsDefined is known to throw System.IO.FileLoadException.
+                // ClassDataContract.IsNonAttributedTypeValidForSerialization is known to throw System.IO.FileNotFoundException.
+                // We guard against all non-critical exceptions.
+            }
+
+            return false;
+        }
+
+        [RequiresUnreferencedCode(DataContract.SerializerTrimmerWarning)]
+        public Type? GetReferencedType(XmlQualifiedName xmlName, DataContract dataContract, out DataContract? referencedContract, out object[]? genericParameters, bool? supportGenericTypes = null)
+        {
+            Type? type = GetReferencedTypeInternal(xmlName, dataContract);
+            referencedContract = null;
+            genericParameters = null;
+
+            if (supportGenericTypes == null)
+                return type;
+
+            if (type != null && !type.IsGenericTypeDefinition && !type.ContainsGenericParameters)
+                return type;
+
+            if (dataContract.GenericInfo == null)
+                return null;
+
+            XmlQualifiedName genericXmlName = dataContract.GenericInfo.GetExpandedXmlName();
+            if (genericXmlName != dataContract.XmlName)
+                throw System.Runtime.Serialization.DiagnosticUtility.ExceptionUtility.ThrowHelperError(new InvalidDataContractException(SR.Format(SR.GenericTypeNameMismatch, dataContract.XmlName.Name, dataContract.XmlName.Namespace, genericXmlName.Name, genericXmlName.Namespace)));
+
+            // This check originally came "here" in the old code. Its tempting to move it up with the GenericInfo check.
+            if (!supportGenericTypes.Value)
```

Comment:
I think you forgot to fill the "here". I'm having a hard time working out the equivalent to the original code and following through the call path. Can you provide a starting method call this method replaces so I can validate the logic? I can find snippets of this code in CodeExporter.GetReferencedType but there's only some of it here, and the parameters don't match at all.

Reply:
I believe "here" meant "here." As in, right where it is in the logical flow of things. The flow of the logic in the old code did the check for `dataContract.GenericInfo == null` and then did the expanded name stuff, and then did this check for `supportGenericTypes`. Not all in a neat single function. This was the crazy refactoring that I had to do to keep a lot of our stuff internal. The comment is just saying that is a simple check of a boolean parameter that in most green-field cases would be checked right at the start. And I am tempted to do just that, but instead left it after name expansion and potential exception throwing.

Anyway, IIRC - and I've fried my brain on a few problems since figuring this out - this refactor/mangle was done to clean up the API between the schema package and the runtime package. The gitst of it is that it was taking logic from a loop that hopped between two classes in the old NetFx that were now in separate packages and we wanted to tighten it up and keep it as much behind one API as we could. The change that did this was part of the "[API cleanup](https://github.com/dotnet/runtime/pull/71752/commits/446c99369d0f14f36f2df1cde45749c762ce2b19#diff-7b5e102cc698e7861e73d2117503765622f232765dd48d17e7621b80b17b104a)" commit. Search that commit for "GenericTypeNameMismatch" to see the various places that the code was and how it migrated into this method.